diff --git a/usr.sbin/jail/config.c b/usr.sbin/jail/config.c
index e6ebc585b..a0e08d905 100644
--- a/usr.sbin/jail/config.c
+++ b/usr.sbin/jail/config.c
@@ -382,8 +382,10 @@ add_param(struct cfjail *j, const struct cfparam *p, enum intparam ipnum,
 			free(dp->name);
 			dp->name = estrdup(name);
 		}
-		if (!(flags & PF_APPEND) || TAILQ_EMPTY(&nss))
+		if (!(flags & (PF_APPEND|PF_PREPEND)) || TAILQ_EMPTY(&nss))
 			free_param_strings(dp);
+		if (0 != (flags & PF_PREPEND))
+			TAILQ_SWAP(&dp->val, &nss, cfstring, tq);
 		TAILQ_CONCAT(&dp->val, &nss, tq);
 		dp->flags |= flags;
 	} else {
diff --git a/usr.sbin/jail/jaillex.l b/usr.sbin/jail/jaillex.l
index cb758b237..3d6d88db4 100644
--- a/usr.sbin/jail/jaillex.l
+++ b/usr.sbin/jail/jaillex.l
@@ -74,9 +74,14 @@ static int lineno = 1;
 			}
 
 			/* Reserved tokens */
-<_>\+=			{
+<_>\[=			{
 				instr = 0;
-				return PLEQ;
+				return OP_PREPEND;
+			}
+<_>\+= |
+<_>\]=			{
+				instr = 0;
+				return OP_APPEND;
 			}
 <_>[,;={}]		{
 				instr = 0;
diff --git a/usr.sbin/jail/jailp.h b/usr.sbin/jail/jailp.h
index 0ed0755f7..039d1fafc 100644
--- a/usr.sbin/jail/jailp.h
+++ b/usr.sbin/jail/jailp.h
@@ -54,6 +54,7 @@
 #define PF_CONV		0x40	/* Parameter duplicated in converted form */
 #define PF_REV		0x80	/* Run commands in reverse order on stopping */
 #define	PF_IMMUTABLE	0x100	/* Immutable parameter */
+#define PF_PREPEND	0x200   /* Prepend to existing parameter list */
 
 #define JF_START	0x0001	/* -c */
 #define JF_SET		0x0002	/* -m */
diff --git a/usr.sbin/jail/jailparse.y b/usr.sbin/jail/jailparse.y
index ff673c726..eac5cd412 100644
--- a/usr.sbin/jail/jailparse.y
+++ b/usr.sbin/jail/jailparse.y
@@ -49,7 +49,8 @@ __FBSDID("$FreeBSD$");
 	char			*cs;
 }
 
-%token      PLEQ
+%token      OP_APPEND
+%token      OP_PREPEND
 %token <cs> STR STR1 VAR VAR1
 
 %type <j>  jail
@@ -117,13 +118,20 @@ param	: name
 		TAILQ_CONCAT(&$$->val, $3, tq);
 		free($3);
 	}
-	| name PLEQ value
+	| name OP_APPEND value
 	{
 		$$ = $1;
 		TAILQ_CONCAT(&$$->val, $3, tq);
 		$$->flags |= PF_APPEND;
 		free($3);
 	}
+	| name OP_PREPEND value
+	{
+		$$ = $1;
+		TAILQ_CONCAT(&$$->val, $3, tq);
+		$$->flags |= PF_PREPEND;
+		free($3);
+	}
 	| name value
 	{
 		$$ = $1;
